File: src/y2016/day19.rs
//! <https://adventofcode.com/2016/day/19>
use crate::{
    framework::{AdventOfCode, ParseError, aoc},
    parser,
};

#[derive(Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct Puzzle {
    input: usize,
}

#[aoc(2016, 19)]
impl AdventOfCode for Puzzle {
    fn parse(&mut self, s: &str) -> Result<(), ParseError> {
        for l in s.lines() {
            self.input = parser::to_usize(l)?;
        }
        Self::parsed()
    }
    fn part1(&mut self) -> Self::Output1 {
        // self.input = 5;
        let mut next = vec![0];
        for i in 1..=self.input {
            next.push(i + 1);
        }
        next[self.input] = 1;
        let mut index = 1;
        while next[index] != index {
            let i = next[next[index]];
            next[index] = i;
            index = i;
        }
        index
    }
    fn part2(&mut self) -> Self::Output2 {
        // self.input = 5;
        let mut next = vec![0];
        for i in 1..=self.input {
            next.push(i + 1);
        }
        next[self.input] = 1;
        let mut remain = self.input;
        let mut index = self.input / 2;
        let mut dist = index;
        while next[index] != index {
            let i = next[next[index]];
            // println!("remain: {remain}, dist: {dist}, {index} directs to {i}");
            next[index] = i;
            remain -= 1;
            if dist == remain / 2 {
                index = i;
            } else {
                dist -= 1;
            }
            debug_assert_eq!(remain / 2, dist);
        }
        index
    }
}

File: swift/Sources/y2016/Day19.swift
//
//  Day19.swift
//
//
//  Created by 楢崎修二 on 2022/06/14.
//

public func day19(_ data: String) {
  let input = 3_014_603
  func part1() {
    var next = [input]
    next.append(contentsOf: Array(2...input + 1))
    next[input] = 1
    var index = 1
    while next[index] != index {
      let i = next[next[index]]
      next[index] = i
      index = i
    }
    print("Part 2: \(index)")
  }
  func part2() {
    var next = [input]
    next.append(contentsOf: Array(2...input + 1))
    next[input] = 1
    var remain = input
    var index = input / 2
    var dist = index
    while next[index] != index {
      let i = next[next[index]]
      next[index] = i
      remain -= 1
      if dist == remain / 2 {
        index = i
      } else {
        dist -= 1
      }
    }
    print("Part 2: \(index)")
  }
  part1()
  part2()
}
